app/controllers/concerns/authentication.rb

```diff
@@ -271,6 +330,14 @@ def request_http_basic_authentication(realm = 'keygen', message = nil)
     raise Keygen::Error::UnauthorizedError.new(code: 'TOKEN_INVALID')
   end
 
+  # NOTE(ezekg) we only support cookie authn from portal origin to prevent CSRF
+  #
+  #             see: https://scotthelme.co.uk/csrf-is-dead/
+  def has_cookie_credentials?
+    (request.origin&.ends_with?(Keygen::Portal::HOST)) &&
```

Comment:
This allows e.g. `evil-safe.com` if portal host were `safe.com`. I think we'd be okay with asserting direct match since we currently don't allow multiple origins (and even then we could introduce `KEYGEN_PORTAL_HOSTS`).